crates/red_knot/src/salsa_db/semantic/types.rs

```diff
@@ -0,0 +1,662 @@
+    positive: FxIndexSet<Type>,
+    // the intersection type does not include any value in any of these types
+    negative: FxIndexSet<Type>,
+}
+
+impl IntersectionType {
+    fn display(
+        &self,
+        f: &mut std::fmt::Formatter<'_>,
+        context: &TypingContext,
+    ) -> std::fmt::Result {
+        f.write_str("(")?;
+        let mut first = true;
+        for (neg, ty) in self
+            .positive
+            .iter()
+            .map(|ty| (false, ty))
+            .chain(self.negative.iter().map(|ty| (true, ty)))
+        {
+            if !first {
+                f.write_str(" & ")?;
+            };
+            first = false;
+            if neg {
+                f.write_str("~")?;
+            };
+            write!(f, "{}", ty.display(context))?;
+        }
+        f.write_str(")")
+    }
+}
+
+pub trait LocalTypeId: Copy {
+    type Ty;
+
+    /// Resolves the type of the data from `types_store`
+    ///
+    /// ## Panics
+    /// May panic if `self` is from another scope than `types_store` or returns an invalid type.
+    fn ty<'a>(&self, types: &'a TypeInference) -> &'a Self::Ty;
+
+    fn intern(ty: Self::Ty, types: &mut TypeInference) -> Self;
+}
+
+pub struct TypingContext<'a> {
+    db: &'a dyn Db,
+    local: Option<(TypingScope, &'a TypeInference)>,
+}
+
+impl<'a> TypingContext<'a> {
+    pub fn local(db: &'a dyn Db, local_scope: TypingScope, types: &'a TypeInference) -> Self {
+        Self {
+            db,
+            local: Some((local_scope, types)),
+        }
+    }
+
+    pub fn global(db: &'a dyn Db) -> Self {
+        Self { db, local: None }
+    }
+
+    pub fn db(&self) -> &'a dyn Db {
+        self.db
+    }
+
+    pub fn types(&self, scope: TypingScope) -> &'a TypeInference {
+        if let Some((local_scope, types)) = self.local {
+            if local_scope == scope {
+                return types;
+            }
+        }
+
+        infer_types(self.db, scope)
+    }
+
+    fn infer_definitions(
+        &self,
+        mut definitions: impl Iterator<Item = ReachableDefinition>,
+        symbol_scope: GlobalId<ScopeId>,
+    ) -> DefinitionType {
+        let Some(first) = definitions.next() else {
+            return DefinitionType::Type(Type::Unknown);
+        };
+
+        let typing_scope = TypingScope::for_symbol_scope(self.db(), symbol_scope);
+        let types = self.types(typing_scope);
+
+        if let Some(second) = definitions.next() {
+            let elements: FxIndexSet<_> = [first, second]
+                .into_iter()
+                .chain(definitions)
+                .map(|definition| self.infer_definition(definition, types))
+                .collect();
+
+            // The fact that the interner is local to a body means that we can't reuse the same union type
+            // across different call sites. But that's something we aren't doing yet anyway. Our interner doesn't
+            // deduplicate union types that are identical.
```

Comment:
Yeah, agree. I think we probably want to have methods on the `TypeInferenceBuilder` because we only need to e.g. track the reverse map of already created unions back to their type ids during construction but we won't need it once type inference is complete (and we won't create any new types)